enhancements/ingress/transition-ingress-from-beta-to-stable.md

```diff
@@ -0,0 +1,302 @@
+
+## Motivation
+
+Some users want to use new features in the `networking.k8s.io/v1` API version,
+and Kubernetes 1.22 will remove the `networking.k8s.io/v1beta1` API version.
+
+The IngressClass API is of particular utility.  With this API, an administrator
+can specify which Ingresses OpenShift should publish and which Ingresses
+OpenShift should ignore, on the assumption that some third-party ingress
+controller publishes them.  This is particularly important with respect to
+status updates, as the Ingress API does not provide a way to differentiate
+status published by one ingress controller from status published by another
+ingress controller, which has the consequence that OpenShift's updates to an
+Ingress's status can interfere with the operation of third-party ingress
+controllers or other controllers that read the Ingress's status.  Thus
+supporting this API is important for accommodating third-party ingress
+controllers in OpenShift clusters.
+
+### Goals
+
+1. Enable users to use new Ingress v1 API features where feasible.
+2. Honor `spec.ingressClassName`.
+3. Prepare for the removal of the v1beta1 API version.
+
+### Non-Goals
+
+1. Extend the Route API to support new Ingress features.
+2. Implement `spec.rules[*].http.paths[*].pathType: Exact`.
+3. Implement new functionality (such as regular expression matching) for `spec.rules[*].http.paths[*].pathType: ImplementationSpecific`.
+4. Implement `spec.defaultBackend`.
+5. Enable Ingresses to use backends that are not Services.
+
+## Proposal
+
+First, the ingress-to-route controller is updated to use the v1 client libraries
+and API types.
+
+Next, the ingress-to-route controller is extended to check for the
+`kubernetes.io/ingress.class` annotation and `spec.ingressClassName` field on
+Ingresses.  If either of these is set, the ingress-to-route controller checks if
+the specified IngressClass exists and, if it does, what its `spec.controller`
+field value is.  If the Ingress specifies an IngressClass that does not exist or
+that does not specify `openshift.io/ingress-to-route` for `spec.controller`, the
+ingress-to-route controller ignores the Ingress.
+
+Additionally, the ingress-to-route controller is extended to check the value of
+`spec.rules[*].http.paths[*].pathType` on Ingresses.  If an Ingress rule
+specifies the value "Exact" for this field, the ingress-to-route controller
+ignores the rule.
+
+Finally, the ingress operator is extended with a new controller that creates an
+IngressClass for each IngressController.  When creating the IngressClass for the
+default IngressController, the operator annotates the IngressClass with the
+`ingressclass.kubernetes.io/is-default-class` annotation if no other
+IngressClass already has that annotation.  Note that the operator does not add
+the annotation when reconciling already created IngressClass objects as the
+administrator may intentionally have configured the cluster to have no default
+IngressClass.
```

Comment:
Perhaps there should be no automatic "is-default-class" annotation if we want to allow the cluster to have no default.  If the admin may intentionally want no default, it doesn't make sense for them to have to go and remove it after we add it.

Reply:
I don't have a clear understanding of why an administrator _wouldn't_ want a default ingressclass; I just didn't want to preclude the possibility, so I added this description of how to configure it.  Having no default ingressclass means that creating an ingress without specifying an ingressclass does not expose the ingress, which is a change in behavior from before this enhancement.  Having the default ingresscontroller's ingressclass be the default one eliminates a configuration step for administrators who want it to be default, which I expect is the most common and least confusing configuration for administrators and users.